homeworks/week7/hw3/script.js

```diff
@@ -0,0 +1,74 @@
+function capitalizeFirst(str) {
+  let result = str[0].toUpperCase()
+  for (let i = 1; i < str.length; i++) {
+    result += str[i]
+  }
+  return result
+}
+document.addEventListener('DOMContentLoaded', () => {
+  // 新增項目
+  const form = document.querySelector('.app__input-box')
+  form.addEventListener('submit', (e) => {
+    e.preventDefault()
+    const div = document.createElement('div')
+    const input = form.querySelector('input')
+    if (!input.value) return
+    const newInput = capitalizeFirst(input.value)
+    div.innerHTML = `
+    <div class="app__todo">
+      <div class="wrapper">
+        <div>
+          <div class="pretty p-default p-round p-thick">
+            <input class="custom-checkbox" type="checkbox" />
+            <div class="state">
+              <label></label>
+            </div>
+          </div>
+        </div>
+        <div class="app__todo__taskname"></div>
+        <form class="edit-form">
+          <input class="app__todo__taskname--edit hide" type="text">
+        </form>
+      </div>
+      <button class="btn-remove">Remove</button>
+    </div>
+`
+    document.querySelector('.app__todos').append(div)
+    const tasknames = document.querySelectorAll('.app__todo__taskname')
+    for (const taskname of tasknames) {
+      if (!taskname.innerText) {
+        taskname.innerText = newInput
+      }
+    }
```

Comment:
我會這樣寫是為了避免用 innerHTML 放 user input，所以才先把 template literals 裡面那一大段 append 上去，後面再用 innerText 的方式把 user input 放到我要的位置。我記得 Huli 在示範影片裡面是用跳脫字的方式來確保 user input 會被解析成 string。 

我現在的解法是在一個 youtube 教學影片上看到的，附上連結給助教參考：
[How To Prevent The Most Common Cross Site Scripting Attack](https://www.youtube.com/watch?v=ns1LX6mEvyM)

Reply:
另外想確認一下，我以為 `:last-child` 取的是某個元素底下的最後一個子元素，可是我的 `.app__todo__taskname` 不是所屬元素的最後一個子元素欸，這樣還可以用 `:last-child` 的方法取到它嗎？我記得當初也試很久取不到，後來才改用迴圈，一圈一圈跑，直到找到那個還沒有加上 innerText 的元素為止。

Reply:
1. 其實大同小異，如果不用老師的跳脫字元方式也是 OK，都是為了避免被解析成 HTML
2. 喔對，我說錯了，應該還是得用 querySelectorAll，但直接取最後一個來改就好 `tasknames[tasknames.length- 1]`，就不用跑回圈

Reply:
對厚我直接抓最後一個就好，完全忘記可以這樣寫！已修正，謝謝 ~